### PR TITLE
Kdf x963 ec changes

### PR DESCRIPTION
**Pull request by slontis**
KDF used by CMS EC Key Agreement has been moved to EVP_KDF. This kdf is almost identical to the the SSKDF hash variant, So it has been implemented in the SSKDF code with its own method table.

##### Checklist

- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

File: crypto/kdf/sskdf.c [hunk 1]
```diff
@@ -70,6 +70,7 @@ static const unsigned char kmac_custom_str[] = { 0x4B, 0x44, 0x46 };
 static int SSKDF_hash_kdm(const EVP_MD *kdf_md,
                           const unsigned char *z, size_t z_len,
                           const unsigned char *info, size_t info_len,
+                          unsigned int append_ctr,
```
Comment (mattcaswell) [inline crypto/kdf/sskdf.c]: Perhaps the comment prior to this function should be updated to additionally refer to X9.63 and explain the `append_ctr` arg

File: crypto/kdf/sskdf.c [hunk 2]
```diff
@@ -468,7 +470,27 @@ static int sskdf_derive(EVP_KDF_IMPL *impl, unsigned char *key, size_t keylen)
...
+
+static int x963kdf_derive(EVP_KDF_IMPL *impl, unsigned char *key, size_t keylen)
+{
+    if (impl->secret == NULL) {
+        KDFerr(KDF_F_X963KDF_DERIVE, KDF_R_MISSING_SECRET);
+        return 0;
+    }
+
+    if (impl->mac != NULL) {
+        return 0;
```
Comment (mattcaswell) [inline crypto/kdf/sskdf.c]: Although other "return 0" paths add an error - but not this one. Shouldn't we have one here too?

File: test/evp_kdf_test.c [hunk 3]
```diff
@@ -174,12 +174,56 @@ static int test_kdf_ss_hash(void)
...
+        0x01, 0x12, 0xb5, 0x85, 0x8f, 0x7a, 0xd3, 0x2a, 0x5a, 0x2d, 0x96, 0xd8,
+        0xcf, 0xfa, 0xbf, 0x7f, 0xa7, 0x33, 0x63, 0x3d, 0x6e, 0x4d, 0xd2, 0xa5,
+        0x99, 0xac, 0xce, 0xb3, 0xea, 0x54, 0xa6, 0x21, 0x7c, 0xe0, 0xb5, 0x0e,
+        0xef, 0x4f, 0x6b, 0x40, 0xa5, 0xc3, 0x02, 0x50, 0xa5, 0xa8, 0xee, 0xee,
+        0x20, 0x80, 0x02, 0x26, 0x70, 0x89, 0xdb, 0xf3, 0x51, 0xf3, 0xf5, 0x02,
+        0x2a, 0xa9, 0x63, 0x8b, 0xf1, 0xee, 0x41, 0x9d, 0xea, 0x9c, 0x4f, 0xf7,
+        0x45, 0xa2, 0x5a, 0xc2, 0x7b, 0xda, 0x33, 0xca, 0x08, 0xbd, 0x56, 0xdd,
+        0x1a, 0x59, 0xb4, 0x10, 0x6c, 0xf2, 0xdb, 0xbc, 0x0a, 0xb2, 0xaa, 0x8e,
+        0x2e, 0xfa, 0x7b, 0x17, 0x90, 0x2d, 0x34, 0x27, 0x69, 0x51, 0xce, 0xcc,
+        0xab, 0x87, 0xf9, 0x66, 0x1c, 0x3e, 0x88, 0x16
```
Comment (mattcaswell) [inline test/evp_kdf_test.c]: Are these self-generated or from a standard? Either way could we have a comment stating the provenance of these values?

File: doc/man7/EVP_KDF_X963.pod [hunk 4]
```diff
@@ -0,0 +1,136 @@
+=pod
+
+=head1 NAME
+
+EVP_KDF_X963 - The X9.63-2001 EVP_KDF implementation
+
+=head1 DESCRIPTION
+
+The EVP_KDF_X963 algorithm implements the key derivation function (X963KDF).
+X963KDF is used by Cryptographic Message Syntax (CMS) for EC KeyAgreement, to
```
Comment (mattcaswell) [inline doc/man7/EVP_KDF_X963.pod]: If this is used in CMS - why don't I see any changes in the CMS code?

Comment (slontis) [reply doc/man7/EVP_KDF_X963.pod]: The CMS entry point is in crypto/ec/ecdh_kdf.c

File: crypto/kdf/sskdf.c (hunk 2 above)
Comment (slontis) [reply crypto/kdf/sskdf.c]: updated

File: crypto/kdf/sskdf.c [hunk 5]
```diff
@@ -66,10 +66,16 @@ static const unsigned char kmac_custom_str[] = { 0x4B, 0x44, 0x46 };
 /*
  * Refer to https://csrc.nist.gov/publications/detail/sp/800-56c/rev-1/final
  * Section 4. One-Step Key Derivation using H(x) = hash(x)
+ * Note: X963 also uses this code with the only difference being that the
```
Comment (mattcaswell) [inline crypto/kdf/sskdf.c]: Shouldn't this really be X9.63 here (i.e. with a dot ".")?

File: crypto/objects/obj_dat.h [hunk 6]
```diff
@@ -2288,9 +2288,10 @@ static const ASN1_OBJECT nid_objs[NUM_NID] = {
     {"SSHKDF", "sshkdf", NID_sshkdf},
     {"SM2-SM3", "SM2-with-SM3", NID_SM2_with_SM3, 8, &so[7766]},
     {"SSKDF", "sskdf", NID_sskdf},
+    {"X963KDF", "x963kdf", NID_x963kdf},
```
Comment (mattcaswell) [inline crypto/objects/obj_dat.h]: I wonder if the name here should have a "dot" too? Not sure

Comment (slontis) [reply crypto/objects/obj_dat.h]: I assume it would be bad in the nid name though :)

Comment (slontis) [reply crypto/objects/obj_dat.h]: maybe use a _ for that case?

Comment (mattcaswell) [reply crypto/objects/obj_dat.h]: The nid name is probably ok as is. Its just whether we add the "." in the sn/ln strings. I lean towards "yes"

Comment (slontis) [reply crypto/objects/obj_dat.h]: I might need to use dash as the nid name comes from the  same fields in objects.txt
# NID for X9.63-2001 KDF
                            : X963KDF            : x963kdf

Comment (mattcaswell) [reply crypto/objects/obj_dat.h]: Hmm. Ok. Not sure which is better, so I'll leave it to you.

Comment (slontis) [reply crypto/objects/obj_dat.h]: I think it will probably be more confusing so I will leave it in a consistent form.

Comment (richsalz) [reply crypto/objects/obj_dat.h]: I think adding the period is not a good thing to do; we always use x509 asn1, etc., and being consistent seems more important.  Folks who care about X9.63 will not be confused by x963.

File: crypto/kdf/sskdf.c (hunk 5 above)
Comment (slontis) [reply crypto/kdf/sskdf.c]: Changed

File: test/evp_kdf_test.c [hunk 7]
```diff
@@ -174,12 +174,61 @@ static int test_kdf_ss_hash(void)
...
+{
+    int ret;
+    EVP_KDF_CTX *kctx = NULL;
+    unsigned char out[1024 / 8];
+    /*
+     * Test data from https://csrc.nist.gov/CSRC/media/Projects/
+     *  Cryptographic-Algorithm-Validation-Program/documents/components/
+     *  800-135testvectors/ansx963_2001.zip
+     */
+    const unsigned char z[] = {
```
Comment (paulidale) [inline test/evp_kdf_test.c]: `static` as well?  Here and below?

Not a critical either way.

Comment (slontis) [reply test/evp_kdf_test.c]: Changed entire file to use statics

**Comment by slontis** [comment]
Updated commit messages

**Comment by paulidale** [comment]
Merged, thanks.

**Pull request closed without merging**